http-client/Network/HTTP/Proxy.hs

```diff
@@ -0,0 +1,355 @@
+import           Control.Arrow               (first)
+import           Control.Monad               (guard)
+import qualified Data.ByteString.Char8       as S8
+import           Data.Char                   (toLower)
+import qualified Data.Map                    as Map
+import qualified Data.Text                   as T
+import           Data.Text.Read              (decimal)
+import           Network.HTTP.Client.Request (applyBasicProxyAuth,
+                                              extractBasicAuthInfo)
+import           Network.HTTP.Client.Types   (HttpExceptionContent (..),
+                                              Proxy (..), Request (..),
+                                              throwHttp)
+import qualified Network.URI                 as U
+import           System.Environment          (getEnvironment)
+
+#if defined(mingw32_HOST_OS)
+import           Control.Exception           (IOException, bracket, catch)
+import           Control.Monad               (join, liftM, mplus, when)
+import           Data.List                   (isPrefixOf)
+import           Foreign                     (Storable (peek, sizeOf), alloca,
+                                              castPtr, toBool)
+import           Network.URI                 (parseAbsoluteURI)
+import           Safe                        (readDef)
+import           System.IO
+import           System.Win32.Registry       (hKEY_CURRENT_USER, rEG_DWORD,
+                                              regCloseKey, regOpenKey,
+                                              regQueryValue, regQueryValueEx)
+import           System.Win32.Types          (DWORD, HKEY)
+#endif
+
+type EnvName     = T.Text
+type HostAddress = S8.ByteString
+type UserName    = S8.ByteString
+type Password    = S8.ByteString
+
+-- There are other proxy protocols like SOCKS, FTP, etc.
+data ProxyProtocol = HTTPProxy | HTTPSProxy
+
+instance Show ProxyProtocol where
+    show HTTPProxy  = "http"
+    show HTTPSProxy = "https"
+
+data ProxySettings = ProxySettings { proxyHost :: Proxy,
+                                     proxyAuth :: Maybe (UserName, Password) }
+                                     deriving Show
+
+httpProtocol :: Bool -> ProxyProtocol
+httpProtocol True  = HTTPSProxy
+httpProtocol False = HTTPProxy
+
+data EnvHelper = EHFromRequest
+               | EHNoProxy
+               | EHUseProxy Proxy
+
+headJust :: [Maybe a] -> Maybe a
+headJust []               = Nothing
+headJust (Nothing:xs)     = headJust xs
+headJust ((y@(Just x)):_) = y
+
+systemProxyHelper :: Maybe T.Text -> ProxyProtocol -> EnvHelper -> IO (Request -> Request)
+systemProxyHelper envOveride prot eh = do
+    let envName' Nothing     = envName prot
+        envName' (Just name) = name
+
+    modifier <- envHelper (envName' envOveride)
+
+-- Under Windows try first env. variables override then Windows proxy settings
+#if defined(mingw32_HOST_OS)
+    modifier' <- systemProxy prot
```

Comment:
It looks like it would be more efficient to check the env vars first and, if that is `Nothing`, _then_ call `systemProxy`. Right now, if `modifier` is `Just`, we'll have an unnecessary `systemProxy` call.